### DISCUSSION TITLE
Breaking change in GLib about empty files

### DISCUSSION THREAD

**Discussion by tsujan** [lxqt/lxqt#2395] [General]
Gnome users should have got used to changes that break their workflows, but such changes usually don't happen to a library as basic as GLib. This time it was GLib's turn:

A recent GLib commit (https://gitlab.gnome.org/GNOME/glib/-/merge_requests/3053) has changed the mimetype of empty files from `text/plain` to `application/x-zerosize`. That has consequences.

`libfm-qt`'s context-menu item "Blank File" opens a dialog with "New text file" as the default text. Previous to the above-mentioned change, a blank file could be opened by any text editor because it was seen as an empty text file. Now, as the MR comment says, "Text editors which want to be opened for new, empty files should add `application/x-zerosize` to their list of supported content types." Most text editors simply don't.

 1. We can leave it as it is. The user could assign `application/x-zerosize` to his/her chosen text editor by using pcmanfm-qt's "Open With..." or  LXQt's File Association dialog. Some text editors might also add support for `application/x-zerosize` (FeatherPad will).
 2. Or we might be able to reverse the change in `libfm-qt` and use `text/plain` for empty files again.

What should we do? Please note that changing "New text file" to "New text file.txt" will have no effect — it will still be seen as `application/x-zerosize`.

**Comment by tsujan** [comment]
Reversing the change in `libfm-qt` is as simple as this:

```diff
diff -ruNp libfm-qt-orig/src/core/fileinfo.cpp libfm-qt/src/core/fileinfo.cpp
--- libfm-qt-orig/src/core/fileinfo.cpp
+++ libfm-qt/src/core/fileinfo.cpp
@@ -53,7 +53,15 @@ void FileInfo::setFromGFileInfo(const GO

     tmp = g_file_info_get_content_type(inf.get());
     if(tmp) {
-        mimeType_ = MimeType::fromName(tmp);
+        if(size_ == 0 && strcmp(tmp, "application/x-zerosize") == 0) {
+            /* Treat zero-sized files as "text/plain",
+               to ensure that they could be opened by any text editor. */
+            mimeType_ = MimeType::fromName("text/plain");
+            icon_ = mimeType_->icon();
+        }
+        else {
+            mimeType_ = MimeType::fromName(tmp);
+        }
     }

     mode_ = g_file_info_get_attribute_uint32(inf.get(), G_FILE_ATTRIBUTE_UNIX_MODE);

```

**Comment by jonesmz** [comment]
As a long time lxqt user, in my opinion the better choice is to adopt the behavior from glib and not undo it. Most distributions will add the mimetype to text editors as they want to, this isn't a big deal.

**Comment by tsujan** [comment]
> in my opinion the better choice is to adopt the behavior from glib and not undo it.

Theoretically, that makes sense to me. But I'm not sure it'll make sense to most LXQt users too: they'll see that, suddenly, the empty files they will create couldn't be opened by their text editors.

**Comment by tsujan** [comment]
This is also a good opportunity to notice a very old fact.

GLib (and so, `libfm-qt`) sees all zero-sized files as one kind of file, but Dolphin (and probably, `kio`) interprets them based on their extensions.

For example, the zero-sized files "empty.txt", "empty.mp3" and "empty.zip" are all seen as the same kind of file with the same icon — previously `text/plain`, now `application/x-zerosize`— while Dolphin sees them as an empty text, audio and zip file, respectively, and gives them their different icons.

We can make use of the current change in GLib and implement Dolphin's behavior.

Currently:

![before](https://user-images.githubusercontent.com/981076/226109017-f724c160-3bc6-46c6-ae26-120a258830c2.png)

After a simple change:

![after](https://user-images.githubusercontent.com/981076/226109028-6b70d0e7-6f37-4df8-a16a-81ecfe564b2b.png)

**Reply by AzumaHazuki** [reply]
Is this wise? That always struck me as a Windows-ism, taking the extension's word for it as to what kind of file is being dealt with. Would it be possible to have some sort of inline "write the corresponding magic numbers to the file headers" logic so that if someone made an empty file and renamed it something.mp3 it would actually get the MP3 file format header written to it?

**Reply by tsujan** [reply]
> That always struck me as a Windows-ism, taking the extension's word for it as to what kind of file is being dealt with.

It isn't so. In Linux, extensions always have priority over contents. Only if there's no (recognizable) extension, the content will be consulted — otherwise, we'd have lots of I/O for no good reason.

> Is this wise? 

Yes! Treating all zero-sized files as having the same type is inconsistent with giving priority to extensions. Instead of fixing this, they just changed the type in https://gitlab.gnome.org/GNOME/glib/-/merge_requests/3053.

The matter has practical aspects too. Besides text editors, apps may be able to open and work with empty files. For example, Arqiver can open "empty.zip" and add files to it (lxqt-archiver can't).

Moreover, when creating files, many apps first create an empty file with an extension and then add content to it. Showing a file with extension as `text/plain` or `application/x-zerosize` in a moment and then changing it to another type with the same extension seems weird.

Fortunately, it can be fixed in `libfm-qt` easily (but not by the above-mentioned patch). IMO, it should have been done a long time ago.

**Comment by tsujan** [comment]
Agreed with @jonesmz and didn't undo https://gitlab.gnome.org/GNOME/glib/-/merge_requests/3053 but enhanced it by giving priority to extensions (like in KDE and Dolphin): https://github.com/lxqt/libfm-qt/pull/891
